compiler/rustc_hir_typeck/src/callee.rs

```diff
@@ -37,9 +37,13 @@ pub(crate) fn check_legal_trait_for_method_call(
     receiver: Option<Span>,
     expr_span: Span,
     trait_id: DefId,
-    _body_id: DefId,
+    body_id: DefId,
 ) -> Result<(), ErrorGuaranteed> {
-    if tcx.is_lang_item(trait_id, LangItem::Drop) {
+    if tcx.is_lang_item(trait_id, LangItem::Drop)
+        // Allow calling `Drop::pin_drop` in `Drop::drop`
+        && let Some(parent) = tcx.opt_parent(body_id)
+        && !tcx.is_lang_item(parent, LangItem::Drop)
```

Comment:
```suggestion
        && !tcx.is_lang_item(tcx.parent(body_id), LangItem::Drop)
```
Body id should have a parent, but if it somehow doesn't, we certainly shouldn't skip the code below.

Reply:
Also, calling the destructor explicitly should probably be unsafe?